### PR TITLE
HipMRI 2D U-Net 43950214

### PR DESCRIPTION
**Pull request by alh1260**
This is my attempt at creating a U-Net for the task of segmenting the HipMRI dataset, implemented in PyTorch. While the code has been proven to function, it currently does not produce very good results, only achieving a Dice similarity coefficient of around 0.3. Perhaps with more time and knowledge it would perform better, however because of pressures from other courses this was unfortunately not the case.

The files included are:
- `report_assets`: this folder contains images used by `README.md`, such as various plots, diagrams and resulting segment maps.
- `comp3710_env.yml`, `comp3710_reduced_env.yml`: these are Conda environment files that reflect the Conda environments I used on the lab computers and on Rangpur, respectively. It is recommended to use these to recreate these environments in order to run the code.
- `dataset.py`: the data loader code for the HipMRI data.
- `dice.py`: the code to calculate the Dice coefficient.
- `hipMri_main.py`: the main driver script.
- The various `hipmri_slices_*.txt` files: these are directory listings of the HipMRI data folders as used by `dataset.py`. These are needed for the data loader to work.
- `modules.py`: the U-Net model code.
- `predict.py`: the code to perform the validation.
- `README.md` and `README.pdf`: the report.
- `testThing.py`: code used to just test various bits of the other code in isolation, as well as used to create predicted segment maps.
- `train.py`: the main training code.

Al Holliday (43950214)

### CODE REVIEW

**Comment by yexincheng** [comment]
**This is an initial inspection, no action is required at this point**
1. **Solves problem**: The average DICE score is around 0.34 on testing set. Detailed introduction of UNet. Loss plot and prediction figures were provided. Proper reference. Good environment setting guide.
2. **Implementation**: Goodimplementation of functions and classes. Hpyerparameters were hard-coded. Using [diceloss](https://github.com/alh1260/PatternAnalysis-2024/blob/topic-recognition/recognition/HipMRI_UNet_Al_Holliday_43950214/predict.py#L38) as the name of a list of dice score  can be confusing.
3. **Design**: Good
4. **Comment**: Sufficient 
5. **Difficulty**: Easy

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Easy. UNet2D (Easy Difficulty)</td><td>-10</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable.  Report tone incorrect. </td><td>-1</td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). Submitted on time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-11</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by hanemma7moud** [comment]
Alastair Holliday

**Comment by shakes76** [comment]
Segmentations not one hot encoded -2 additionally, low quality results -1
